cruise-control-client/cruisecontrolclient/client/Responder.py

```diff
@@ -0,0 +1,207 @@
+    """
+
+    def __init__(self, url: str, headers: Dict[str, str] = None):
+        if 'json=true' not in url:
+            raise ValueError("url must contain the \"json=true\" parameter")
+        AbstractResponder.__init__(self, url, headers)
+        # This abstract class _still_ does not define which HTTP method should be used
+
+    def retrieve_response(self) -> requests.Response:
+        """
+        Returns a final requests.Response object from cruise-control
+        where Response.text is JSON-formatted.
+
+        :return: requests.Response
+        """
+        # cruise-control's JSON response has a 'progress' key in it so long
+        # as the response is not final.
+        #
+        # Once the response is final, it does not contain the 'progress' key.
+        #
+        # Accordingly, keep getting the response from this session and checking
+        # it for the 'progress' key.
+        #
+        # Return the response as JSON once we get a valid JSON response that we
+        # think is 'final'.
+
+        # Alert the humans to long-running poll
+        if self.print_to_stdout_enabled:
+            print_error("Starting long-running poll of {}".format(self.url))
+
+        # TODO: Session.get and Session.post can return a plethora of exceptions;
+        # they should be handled here.
+        response = self.session_http_method(self.url, headers=self.headers)
+        while 'progress' in response.json().keys():
+            if self.print_to_stdout_enabled:
+                display_response(response)
+            response = self.session_http_method(self.url, headers=self.headers)
+
+        # return the requests.response object
+        return response
+
+
+class AbstractJSONDisplayingResponder(AbstractJSONResponder):
+    print_to_stdout_enabled = True
+    """
+    This class displays intermediate responses to humans via stdout.
+    """
+
+
+class JSONDisplayingResponderGet(AbstractJSONDisplayingResponder):
+    """
+    This class returns a final requests.Response object from cruise-control
+    where Response.text is JSON-formatted, and where the HTTP method is GET.
+    """
+
+    def __init__(self, url: str, headers: Dict[str, str] = None):
+        AbstractJSONDisplayingResponder.__init__(self, url, headers)
+        self.session_http_method = self.session.get
+
+
+class JSONDisplayingResponderPost(AbstractJSONDisplayingResponder):
+    """
+    This class returns a final requests.Response object from cruise-control
+    where Response.text is JSON-formatted, and where the HTTP method is POST.
+    """
+
+    def __init__(self, url: str, headers: Dict[str, str] = None):
+        AbstractJSONDisplayingResponder.__init__(self, url, headers)
+        self.session_http_method = self.session.post
+
+
+class AbstractJSONResponderThread(Thread):
+    """
+    This abstract class defines a Thread whose purpose is to communicate with
+    cruise-control and return a requests.Response object.
```

Comment:
Nit (multiple occurrences): `a requests` -> `a request`

Reply:
This docstring refers to the `requests.Response` object that is returned by the thread.

Reply:
Ah, my bad :)